extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/ReadMongoP.java

```diff
@@ -0,0 +1,472 @@
+ * Processor for reading from MongoDB
+ *
+ * Reading is done by one of two readers:
+ * <ul>
+ * <li>batch reader, which uses {@linkplain MongoCollection#aggregate} to find matching documents.</li>
+ * <li>streaming reader, which uses {@linkplain MongoClient#watch} (or same function on database or collection level)
+ * to find matching documents, as they arrive in the stream.</li>
+ *
+ * All processing guarantees are supported via standard snapshotting mechanism. Each instance of this processor
+ * will save it's state (last read key or resumeToken) with the key being global processor index mod total parallelism.
+ * </ul>
+ * @param <I> type of emitted item
+ */
+public class ReadMongoP<I> extends AbstractProcessor {
+
+    private static final int BATCH_SIZE = 1000;
+    private ILogger logger;
+
+    private int totalParallelism;
+    private int processorIndex;
+
+    private boolean snapshotsEnabled;
+
+    private boolean snapshotInProgress;
+    private final MongoChunkedReader reader;
+    private final MongoDbConnection connection;
+
+    private Traverser<?> traverser;
+    private Traverser<? extends Entry<BroadcastKey<Integer>, ?>> snapshotTraverser;
+
+    /**
+     * Creates a new processor with batch reader.
+     */
+    public ReadMongoP(
+            SupplierEx<? extends MongoClient> clientSupplier,
+            List<Bson> aggregates,
+            String databaseName,
+            String collectionName,
+            FunctionEx<Document, I> mapItemFn
+    ) {
+        this.reader = new BatchMongoReader(databaseName, collectionName, mapItemFn, aggregates);
+        this.connection = new MongoDbConnection(clientSupplier, client -> reader.connect(client, snapshotsEnabled));
+    }
+
+    /**
+     * Creates a new processor with streaming reader.
+     */
+    public ReadMongoP(
+            SupplierEx<? extends MongoClient> clientSupplier,
+            Long startAtTimestamp,
+            EventTimePolicy<? super I> eventTimePolicy,
+            List<Bson> aggregates,
+            String databaseName,
+            String collectionName,
+            FunctionEx<ChangeStreamDocument<Document>, I> mapStreamFn
+    ) {
+        EventTimeMapper<I> eventTimeMapper = new EventTimeMapper<>(eventTimePolicy);
+        eventTimeMapper.addPartitions(1);
+        this.reader = new StreamMongoReader(databaseName, collectionName, mapStreamFn,
+                startAtTimestamp, aggregates, eventTimeMapper);
+        this.connection = new MongoDbConnection(clientSupplier, client -> reader.connect(client, snapshotsEnabled));
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        logger = context.logger();
+        totalParallelism = context.totalParallelism();
+        processorIndex = context.globalProcessorIndex();
+        this.snapshotsEnabled = context.snapshottingEnabled();
+
+        connection.reconnectIfNecessary();
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
```

Comment:
Could you add a comment that the async driver is not developed anymore, so we don't have a simple way to make this cooperative at this moment.